### ISSUE TITLE
Test & potentially provide Linux & Windows compatibility

### ISSUE THREAD

**Issue by mattstauffer** [tighten/takeout#37]

**Comment by jyrkidn** [comment]
Hey, I'm working daily on Ubuntu and eager to test this out. I'm going to make a fork and see where I end up. Since I'm already stuck on this line https://github.com/tightenco/takeout/blob/fc068703fc6423ca880cc2ce7c2a85d81b7589b5/app/Shell/Docker.php#L113

**Comment by jyrkidn** [comment]
Made some progress already, this was the only thing I had to change: https://github.com/jyrkidn/takeout/commit/aea9b092a114dd5b8944473e6d5c1fcd6a9a8b91

And I was able to enable the mysql service. I'll test further in the next days/weeks with other services.

**Comment by armancodes** [comment]
@mattstauffer It would be great if Takeout will be available on Windows.

**Comment by mattstauffer** [comment]
Thanks!! I actually have a PC laptop in the mail that I plan to use to test it on Linux and windows!

**Comment by PHLAK** [comment]
@mattstauffer Do you have any idea offhand of what needs to be done to support Linux? I may look into this and having a place to start would be helpful.

**Comment by mattstauffer** [comment]
@PHLAK Hopefully very little. My guess is the main things will be:

1. The check that ensures Docker is installed and running
2. The port checks

**Comment by PHLAK** [comment]
I just did a quick test and you're right @mattstauffer, the check to ensure Docker is running fails. The problem is in `App\Shell\Docker::isDockerServiceRunning()` where it does 

```php
$this->shell->execQuietly('launchctl list | grep com.docker.docker');
```

If I change this to 

```php
$this->shell->execQuietly('docker info');
````

it works as expected and I was able to get a container up and running.

I believe 'docker info' should work on Mac but don't have one to verify. If someone is able to verify this I'd be happy to submit a PR with the fix (or go ahead and do it yourself).

```
$ ./takeout enable

<picked elastic search from the menu>

Which host port would you like this service to use? [9200]:
 > 

 Which tag (version) of this service would you like to use? [latest]:
 > 

 What is the Docker volume name? [elastic_data]:
 > 

Downloading docker image...

 OUT  7.9.0: Pulling from library/elasticsearch
 OUT  75f829a71a1c: Pulling fs layer
372e2d4c0ddd: Pulling fs layer
8198cfd92e19: Pulling fs layer
2f0fcc2d3406: Pulling fs layer

<additional output truncated>

 OUT  6a35898836f0: Pull complete
 OUT  7d6737d49278: Pull complete
 OUT  Digest: sha256:e72555746531240fcc1d8eb95866803bf6b6b91f46f1bc4d573ebdd9fcd951ba
 OUT  Status: Downloaded newer image for elasticsearch:7.9.0
 OUT  docker.io/library/elasticsearch:7.9.0
Enabling elasticsearch...

 OUT  032c6771e58bf0ad1d74abb88460374bfce22891364a1c9fd43494352e9ee764

Service enabled!

$ docker ps -a
CONTAINER ID        IMAGE                   COMMAND                  CREATED             STATUS                  PORTS                               NAMES
032c6771e58b        elasticsearch:7.9.0     "/tini -- /usr/local…"   6 minutes ago       Up 6 minutes            0.0.0.0:9200->9200/tcp, 9300/tcp    TO--elasticsearch--7.9.0
```

I'll try to check out the port stuff later.

**Comment by mattstauffer** [comment]
@PHLAK Thanks! I verified and it worked, so I put up a PR with you as co-author... gotta share that credit! Thanks so much!

**Comment by mattstauffer** [comment]
Looks like we also have a PR to add Windows compatibility for that check! I'll refactor, test, and merge that once #100 is merged.

https://github.com/tightenco/takeout/pull/93/files

**Comment by PHLAK** [comment]
I just checked and `docker info` also works as expected on Windows.

**Comment by sinnbeck** [comment]
I am working on a laravel package inspired by Takeout and have gotten it to work on all 3 platforms. To get it working on windows you need to reformat all your command strings from `"$variable" to `"${:$variable}"`

This will make it work correctly on all 3 platforms

For example for mysql, it should be
```
protected $dockerRunTemplate = '-p "${:port}":3306 \
        -e MYSQL_ROOT_PASSWORD="${:root_password}" \
        -v "${:volume}":/var/lib/mysql \
        "${:organization}"/"${:image_name}":"${:tag}"';
```
https://symfony.com/doc/current/components/process.html#using-features-from-the-os-shell

**Comment by mattstauffer** [comment]
@sinnbeck Great--thanks so much for your help!

**Comment by jonsugar** [comment]
Hey @josecanhelp, 
cc: @mattstauffer

I recently spoke with Matt in DM about takeout and OS support to query why requirements state MacOS since I'm using it quite happily in an Ubuntu 20.04 instance under WSL2 with Windows docker. Matt mentioned he'd like to update requirements if WSL is indeed working.

**TL;DR** How can I assist in validating WSL2+Windows docker as a supported environment. 

**My Set-up**
| Dependency | Version| Notes
|---|---|---
| Windows 10 | 2004 | minimum required version for WSL2
| Docker Desktop | 2.3.0.5 stable | minimum required version 2.3.0.2 stable
| Ubuntu | 20.04.1 LTS | installed via the Windows store
|takeout| 1.4.0 | --

To make it work I have enabled WSL2 integration in windows docker. My best understanding is this uses smoke and mirrors to proxy windows docker as **/usr/bin/docker** within the Linux distro. See [Docker Desktop WSL 2 backend](https://docs.docker.com/docker-for-windows/wsl/).

I searched the takeout issues and found this issue and #112 and #57. I don't think it makes sense to shut down the docker service from within the WSL instance. Takeout will need to identify it is running in a Windows WSL Linux environment. This can be determined using **uname -r** which reports the kernel release.

```bash
# determine we are running in a WSL container
uname -r
4.19.104-microsoft-standard
```

A quick search in the takeout codebase reveals that stopping docker after last service shutdown is not applied to Linux OSes:

```php
if (count($this->docker->allContainers()) === 0 && in_array(PHP_OS_FAMILY, ['Darwin','Windows'])) {
   ⁝
        $this->task('Stopping Docker service ', $this->docker->stopDockerService());
   ⁝ 
}
```

So given all the above (notwithstanding anything else you are aware off), I think that Windows+WSL2+Windows Docker is currently stable  🤷

**Comment by delta1186** [comment]
@mattstauffer so I am working on a PR to add support for Windows (not WSL).  One of the first issues I am facing is the fact that the nunomaduro/laravel-console-menu package does not work on Windows due to the lack of support for the ext_posix php extension.  I ran into this when running composer install.  So actually two questions.  

First, would you be opposed to adding the following to the composer.json file?  This I believe this should prevent Windows users from having to add the `--ignore-platform-reqs` when running composer install while developing on takeout.

`"config": {
        "platform": {
            "ext-pcntl": "7.2",
            "ext-posix": "7.2"
        }
    },`

Second, concerning the menu code for Windows.  Would something like the following be acceptable?  This is an example from EnableCommand.


```
 if (! in_array(PHP_OS_FAMILY, ['Windows'])) {
            $option = $this->menu('Services to enable', $this->enableableServices())
                ->addLineBreak('', 1)
                ->setPadding(2, 5)
                ->open();
        }
        if (in_array(PHP_OS_FAMILY, ['Windows'])) {
            $option = $this->choice('What service would you like to enable?', $this->enableableServices());
        }
```
I just want to get your thoughts before I venture down the rabbit hole.  Thanks!

**Comment by mattstauffer** [comment]
@jonsugar Thanks so much for testing it! So, if I'm reading you right, it's 100% WSL-compatible since, I assume, `PHP_OS_FAMILY` would come through as Linux? If so, I'll mark it as WSL-compatible, which would suggest it's also possibly fully Linux compatible at this point?

**Comment by mattstauffer** [comment]
@delta1186 Would you mind opening this comment up as a separate issue? I'm inclined to say "yes" to both of your concerns, but I'd like to make sure we consider all of the options and concerns. Thanks!

**Comment by delta1186** [comment]
@mattstauffer sure thing. I'll do it in the AM. Gotten finish watching this train wreck. 🤦🏻‍♂️

**Comment by delta1186** [comment]
@mattstauffer sure thing! I'll make it happen first thing tomorrow. Thanks for the feed back.

**Comment by jonsugar** [comment]
@mattstauffer 

>If I'm reading you right, it's 100% WSL-compatible since, I assume, PHP_OS_FAMILY would come through as Linux? 

Yes, with the caveat that my testing is for **WSL2**. My experience with the original WSL wasn't the best so I never used it in anger for dev. WSL2 seems faster and just generally more seamless. 

>If so, I'll mark it as WSL-compatible, which would suggest it's also possibly fully Linux compatible at this point?

Yeah, I think so.

**Context** 

Maybe the following will help give some context as to my experience with WSL vs WSL2. 

From, from https://docs.docker.com/docker-for-windows/wsl/

>Windows Subsystem for Linux (WSL) 2 introduces a significant architectural change as it is a full Linux kernel built by Microsoft, allowing Linux containers to run natively without emulation. With Docker Desktop running on WSL 2, users can leverage Linux workspaces and avoid having to maintain both Linux and Windows build scripts. In addition, WSL 2 provides improvements to file system sharing, boot time, and allows access to some cool new features for Docker Desktop users.

>Docker Desktop uses the dynamic memory allocation feature in WSL 2 to greatly improve the resource consumption. This means, Docker Desktop only uses the required amount of CPU and memory resources it needs, while enabling CPU and memory-intensive tasks such as building a container to run much faster.

>Additionally, with WSL 2, the time required to start a Docker daemon after a cold start is significantly faster. It takes less than 10 seconds to start the Docker daemon when compared to almost a minute in the previous version of Docker Desktop.

**Comment by mattstauffer** [comment]
I'm going to close this issue and replace it with separate issues for each operating system, because it seems we're there-or-close for Linux and WSL2. 🎉
